File: Olympiad/POI/official/04-XI/pin/hsm_zaw.pas
{
  Implementacja przyk³adowego modulu HSM dla zawodnika w Pascalu
  Autor: Tomek Czajka
}

unit hsm;

interface

function sprawdz(pin: array of Longint; a: array of Longint): Boolean;

procedure wynik(pin: array of Longint);


implementation

const zainicjowany: Boolean = false;
const zapytania: Longint = 0;
var zaszyfrowany: array[0..3] of Longint;

procedure inicjuj;
var
  buf: String;
  i,x: Longint;
  
begin
  write('Podaj zaszyfrowany numer karty (szesnastkowo): ');
  read(buf);
  if length(buf)<4 then begin
    writeln('Za krotki');
    halt(1);
  end;
  for i:=0 to 3 do begin
    if (buf[i+1]>='0') and (buf[i+1]<='9') then x:=ord(buf[i+1])-ord('0')
    else if (buf[i+1]>='a') and (buf[i+1]<='f') then x:=ord(buf[i+1])-ord('a')+10
    else if (buf[i+1]>='A') and (buf[i+1]<='F') then x:=ord(buf[i+1])-ord('A')+10
    else begin
      writeln('Zla litera');
      halt(1);
    end;
    zaszyfrowany[i] := x;
  end;
  zainicjowany := true;
end;

function sprawdz(pin: array of Longint; a: array of Longint): Boolean;
var i: Longint;
begin
  if not zainicjowany then inicjuj();
  
  inc(zapytania);

  for i:=0 to 3 do begin
    if (pin[i]<0) or (pin[i]>9) then begin
      writeln('PIN spoza zakresu');
      halt(1)
    end
  end;
  
  for i:=0 to 15 do begin
    if (a[i]<0) or (a[i]>9) then begin
      writeln('element tablicy konwersji spoza zakresu');
      halt(1)
    end
  end;

  for i:=0 to 3 do begin
    if pin[i]<>a[zaszyfrowany[i]] then exit(false);
  end;
  
  exit(true);
end;

procedure wynik(pin: array of Longint);
var
  i: Longint;
  a: array [0..15] of Longint;
begin
  for i:=0 to 15 do a[i]:=i mod 10;
  
  if sprawdz(pin,a) then writeln('OK')
  else writeln('ZLE');
  
  dec(zapytania);
  
  writeln('Liczba zapytan: ', zapytania);
  halt(0);
end;

begin
end.

File: Olympiad/POI/official/2004/code/pin/hsm_zaw.c
/*
  Implementacja przyk³adowego modulu HSM dla zawodnika w C/C++
  Autor: Tomek Czajka
*/

#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#include "hsm.h"

static int zainicjowany = 0;

static int zaszyfrowany[4];
static int zapytania = 0;

static void inicjuj() {
  int i,x;
  char buf[100];
  
  printf("Podaj zaszyfrowany numer karty (szesnastkowo): ");
  scanf("%s",buf);
  if(strlen(buf)<4) {
    printf("Za krotki\n");
    exit(1);
  }
  for(i=0;i<4;++i) {
    if(buf[i]>='0' && buf[i]<='9') x=buf[i]-'0';
    else if(buf[i]>='a' && buf[i]<='f') x=buf[i]-'a'+10;
    else if(buf[i]>='A' && buf[i]<='F') x=buf[i]-'A'+10;
    else {
      printf("Zla litera\n");
      exit(1);
    }
    zaszyfrowany[i] = x;
  }
  zainicjowany = 1;
}

int sprawdz(int p[], int a[]) {
  int i;
  
  if(!zainicjowany) inicjuj();
  
  ++zapytania;

  for(i=0;i<4;++i) {
    if(p[i]<0 || p[i]>9) {
      printf("PIN spoza zakresu\n");
      exit(1);
    }
  }
  for(i=0;i<16;++i) {
    if(a[i]<0 || a[i]>9) {
      printf("element tablicy konwersji spoza zakresu\n");
      exit(1);
    }
  }
  
  for(i=0;i<4;++i) {
    if(p[i] != a[zaszyfrowany[i]])
      return 0;
  }
  return 1;
}

void wynik(int p[]) {
  int i;
  int a[16];
  
  for(i=0;i<16;++i) a[i]=i%10;
  if(sprawdz(p,a)) printf("OK\n");
  else printf("ZLE\n");
  --zapytania;
  printf("Liczba zapytan: %d\n", zapytania);
  exit(0);
}
